sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java

```diff
@@ -0,0 +1,125 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.ddl.index;
+
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.support.master.AcknowledgedRequest;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+public class BulkRenameIndexRequest extends AcknowledgedRequest<BulkRenameIndexRequest> {
+
+    public class RenameIndexAction {
+
+        private final String sourceIndexName;
+        private final String targetIndexName;
+
+        public RenameIndexAction(@Nonnull String sourceIndexName,
+                                 @Nonnull String targetIndexName) {
+            this.sourceIndexName = Objects.requireNonNull(sourceIndexName,
+                "Source Index Name must not be null");
+            this.targetIndexName = Objects.requireNonNull(targetIndexName,
+                "Target Index Name must not be null");
+            if (sourceIndexName.isEmpty() || targetIndexName.isEmpty()) {
+                throw new IllegalStateException("Source or Target Index Names must not be empty");
+            }
+        }
+
+        @Nonnull String sourceIndexName() {
+            return sourceIndexName;
+        }
+
+        @Nonnull String targetIndexName() {
+            return targetIndexName;
+        }
+    }
+
+    private List<RenameIndexAction> renameIndexActions;
+
+    public BulkRenameIndexRequest(int initialNumberOfRenameRequests) {
+        this.renameIndexActions = new ArrayList<>(initialNumberOfRenameRequests);
+    }
+
+    BulkRenameIndexRequest() {
+        this(3);
+    }
+
+    public void addRenameIndexAction(@Nonnull String sourceIndex,
+                                     @Nonnull String targetIndex) {
+        renameIndexActions.add(new RenameIndexAction(sourceIndex, targetIndex));
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (renameIndexActions.isEmpty()) {
+            validationException = addValidationError("Rename Index Action list must not be empty", null);
+        }
+        return validationException;
+    }
+
+    public List<RenameIndexAction> renameIndexActions() {
+        return renameIndexActions;
+    }
+
+    public String[] sourceIndices() {
```

Comment:
I think we should remove these here and have the call-site execute this logic if it is required. Or better have them use the `RenameIndexAction` structure directly. If we can use the type system to enforce a pairing of source and target we should do it, and not introduce a loose contract that two independent arrays must fit together.

Reply:
Also something I had in place and changed it to be more in line of the existing APIs that operate on Arrays. I will change it